File: apps/mobile/src/lib/money.ts
const ZERO = new Set(['JPY', 'KRW', 'VND', 'CLP', 'ISK', 'IDR', 'HUF', 'TWD', 'UGX', 'XAF', 'XOF']);
const THREE = new Set(['KWD', 'BHD', 'JOD', 'OMR', 'TND', 'LYD', 'IQD']);

export function decimals(currency: string): number {
  if (ZERO.has(currency)) return 0;
  if (THREE.has(currency)) return 3;
  return 2;
}

export function toMinor(input: string, currency: string): number | null {
  const clean = input.replace(/[\s,]/g, '');
  if (!/^\d+(\.\d+)?$/.test(clean)) return null;
  return Math.round(Number(clean) * 10 ** decimals(currency));
}

export function format(minor: number, currency: string): string {
  const d = decimals(currency);
  return new Intl.NumberFormat(undefined, {
    style: 'currency',
    currency,
    minimumFractionDigits: d,
    maximumFractionDigits: d,
  }).format(minor / 10 ** d);
}

// rates are "units of X per one unit of home", as returned by /v1/fx?base=home
export function toHome(minor: number, from: string, home: string, rates: Record<string, number>): number | null {
  if (from === home) return minor;
  const rate = rates[from];
  if (!rate) return null;
  const major = minor / 10 ** decimals(from) / rate;
  return Math.round(major * 10 ** decimals(home));
}

export const THRESHOLDS = [50, 80, 100] as const;

export function crossed(beforeMinor: number, afterMinor: number, budgetMinor: number): number | null {
  if (budgetMinor <= 0) return null;
  const pct = (v: number) => (v / budgetMinor) * 100;
  const hit = THRESHOLDS.filter((t) => pct(beforeMinor) < t && pct(afterMinor) >= t);
  return hit.length ? hit[hit.length - 1] : null;
}

File: services/api/src/money.rs
use serde::{Deserialize, Serialize};

/// ISO 4217 code, uppercased.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(try_from = "String", into = "String")]
pub struct Currency(String);

impl Currency {
    pub fn parse(raw: &str) -> Option<Self> {
        let code = raw.trim().to_ascii_uppercase();
        (code.len() == 3 && code.bytes().all(|b| b.is_ascii_uppercase())).then_some(Self(code))
    }

    pub fn exponent(&self) -> u32 {
        match self.0.as_str() {
            "BIF" | "CLP" | "DJF" | "GNF" | "ISK" | "JPY" | "KMF" | "KRW" | "PYG" | "RWF"
            | "UGX" | "VND" | "VUV" | "XAF" | "XOF" | "XPF" => 0,
            "BHD" | "IQD" | "JOD" | "KWD" | "LYD" | "OMR" | "TND" => 3,
            _ => 2,
        }
    }
}

impl TryFrom<String> for Currency {
    type Error = String;

    fn try_from(raw: String) -> Result<Self, Self::Error> {
        Self::parse(&raw).ok_or_else(|| format!("invalid currency {raw:?}"))
    }
}

impl From<Currency> for String {
    fn from(c: Currency) -> Self {
        c.0
    }
}

/// Parses a printed amount ("1,234.50", "฿ 120", "1.234,50") into integer minor units.
pub fn to_minor(printed: &str, currency: &Currency) -> Option<i64> {
    if printed.contains('-') {
        return None;
    }
    let kept: String = printed
        .chars()
        .filter(|c| c.is_ascii_digit() || *c == '.' || *c == ',')
        .collect();
    let exp = currency.exponent();
    let (whole, frac) = match kept.rfind(['.', ',']) {
        Some(p) => {
            let frac = &kept[p + 1..];
            if (1..=exp as usize).contains(&frac.len()) {
                (&kept[..p], frac)
            } else if frac.len() == 3 {
                (kept.as_str(), "")
            } else {
                return None;
            }
        }
        None => (kept.as_str(), ""),
    };
    let whole: String = whole.chars().filter(char::is_ascii_digit).collect();
    if whole.is_empty() && frac.is_empty() {
        return None;
    }
    let units: i64 = if whole.is_empty() {
        0
    } else {
        whole.parse().ok()?
    };
    let frac: i64 = format!("{frac:0<width$}", width = exp as usize)
        .parse()
        .unwrap_or(0);
    units.checked_mul(10_i64.pow(exp))?.checked_add(frac)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn c(code: &str) -> Currency {
        Currency::parse(code).unwrap()
    }

    #[test]
    fn converts_printed_amounts_to_minor_units() {
        let cases = [
            ("plain", "12.50", "AUD", Some(1250)),
            ("thousands comma", "1,234.50", "USD", Some(123_450)),
            ("european format", "1.234,50", "EUR", Some(123_450)),
            ("decimal comma", "12,5", "EUR", Some(1250)),
            ("symbol and space", "฿ 120", "THB", Some(12_000)),
            ("zero-decimal currency", "1,500", "JPY", Some(1500)),
            (
                "zero-decimal with grouping dots",
                "150.000",
                "VND",
                Some(150_000),
            ),
            ("three-decimal currency", "1.250", "KWD", Some(1250)),
            ("more decimals than currency allows", "12.3456", "USD", None),
            (
                "three digits after separator is grouping",
                "12.345",
                "INR",
                Some(1_234_500),
            ),
            ("empty", "", "USD", None),
            ("negative", "-5.00", "USD", None),
            ("garbage", "abc", "USD", None),
        ];
        for (name, printed, cur, want) in cases {
            assert_eq!(to_minor(printed, &c(cur)), want, "case {name}");
        }
    }

    #[test]
    fn exponent_defaults_to_two() {
        assert_eq!(c("jpy").exponent(), 0);
        assert_eq!(c("KWD").exponent(), 3);
        assert_eq!(c("INR").exponent(), 2);
        assert!(Currency::parse("EURO").is_none());
    }
}
